### PR TITLE
[Refactor] Update RNG code from upstream

### PR DESCRIPTION
**Pull request by Fuzzbawls**
This is a collection of upstream PRs that have been backported to bring our RNG (`src/random`) code more up-to-date. The following upstream PRs have been included here:

- https://github.com/bitcoin/bitcoin/pull/12742
- https://github.com/bitcoin/bitcoin/pull/14624 
  - some of this had already been merged previously
- https://github.com/bitcoin/bitcoin/pull/14955
- https://github.com/bitcoin/bitcoin/pull/15250
- https://github.com/bitcoin/bitcoin/pull/15224
- https://github.com/bitcoin/bitcoin/pull/15324
- https://github.com/bitcoin/bitcoin/pull/15296
- https://github.com/bitcoin/bitcoin/pull/15512
- https://github.com/bitcoin/bitcoin/pull/16878
- https://github.com/bitcoin/bitcoin/pull/17151
- https://github.com/bitcoin/bitcoin/pull/17191
- https://github.com/bitcoin/bitcoin/pull/13236
- https://github.com/bitcoin/bitcoin/pull/13314
- https://github.com/bitcoin/bitcoin/pull/17169
- https://github.com/bitcoin/bitcoin/pull/17270
  -  omitted last commit as our testing framework doesn't support it currently
  - omitted https://github.com/bitcoin/bitcoin/pull/17270/commits/64e1e022cedf6776c5dffd488ca2e766adca5dc3, to be pulled in after our time utility is updated in a separate PR
- https://github.com/bitcoin/bitcoin/pull/17573
- https://github.com/bitcoin/bitcoin/pull/17507
- https://github.com/bitcoin/bitcoin/pull/17670
- https://github.com/bitcoin/bitcoin/pull/17527
- https://github.com/bitcoin/bitcoin/pull/14127
- https://github.com/bitcoin/bitcoin/pull/21486

### CODE REVIEW

**Review by random-zebra — APPROVED** [review]
Really good backport. ACK 77c5f58477bacc169da959dad0bb96a665eaba0d

File: src/libzerocoin/bignum_gmp.cpp [hunk 1]
```diff
@@ -63,7 +63,7 @@ CBigNum CBigNum::randBignum(const CBigNum& range)
     size_t size = (mpz_sizeinbase (range.bn, 2) + CHAR_BIT-1) / CHAR_BIT;
     std::vector<unsigned char> buf(size);
 
-    RandAddSeed();
+    RandAddSeedSleep();
     GetRandBytes(buf.data(), size);
```
[the remark this answers is not in the document]
Comment (Fuzzbawls) [reply src/libzerocoin/bignum_gmp.cpp]: changed it, and yeah...most of this isn't even used anymore

**Comment by Fuzzbawls** [comment]
Added 2 more PRs to this to detect and guard the `ifaddrs.h` header file detection, which is used as an additional entropy source for randomness

**Review by furszy — APPROVED** [review]
pretty cool👌, we are close to latest upstream with this changes.
ACK cecbf6c with a minor nit that can be easily tackled later.

**Review by random-zebra — APPROVED** [review]
rebase utACK cecbf6c0ae4a70dd05390bd6979f6f2cf893158f and merging...

**Pull request merged**